design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,809 @@
+Ginkgo.
+
+TestMain:
+
+```go
+var testenv *env.Environment
+
+func TestMain(m *testing.M) {
+	testenv = env.NewInClusterConfig()
+
+	testenv.Setup(
+		setupSchema,
+		setupDummyProvider,
+	)
+
+	testenv.Finish(
+		teardownDummyProvider,
+	)
+
+	// launch package tests
+	os.Exit(testenv.Run(m))
+}
+```
+
+The tests themselves are organized in Features for which steps (Setup, Assess,
+Teardown) need to be defined.
+
+Test example:
+
+```go
+func TestFlow(t *testing.T) {
+
+	f := features.New("create claim flow").
+		WithLabel("type", "flow-claim").
+		WithSetup("install the XRD", setupXRD).
+		WithTeardown("teardown the XRD", teardownXRD).
+		WithSetup("install the Composition", setupComposition).
+		WithTeardown("teardown the Composition", teardownComposition).
+		WithSetup("xrd is established and offered", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			xrd := genXRD()
+			err := wait.For(conditions.New(cfg.Client().Resources()).ResourceMatch(xrd, func(object k8s.Object) bool {
+				o := object.(*extv1.CompositeResourceDefinition)
+				return o.Status.GetCondition(extv1.TypeEstablished).Status == corev1.ConditionTrue &&
+					o.Status.GetCondition(extv1.TypeOffered).Status == corev1.ConditionTrue
+			}), wait.WithTimeout(time.Minute*1))
+			if err != nil {
+				t.Fatalf("failed to wait for XRD to be established and offered: %v", err)
+			}
+			return ctx
+		}).
+		Setup(setupClaim).
+		Teardown(teardownClaim).
+		Assess("claim", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			claim, err := getClaim()
+			if err != nil {
+				t.Fatalf("failed to get claim: %v", err)
+			}
+			err = wait.For(conditions.New(cfg.Client().Resources()).ResourceMatch(claim, func(object k8s.Object) bool {
+				claimObj := composed.Unstructured{Unstructured: *claim}
+				isReady := claimObj.GetCondition(xpv1.TypeReady)
+				if isReady.Status != corev1.ConditionTrue {
+					t.Logf("claim %q is not yet Ready", claim.GetName())
+					return false
+				}
+				return isReady.Status == corev1.ConditionTrue
+			}), wait.WithTimeout(time.Minute*1))
+			if err != nil {
+				t.Fatalf("failed to wait for claim to be Ready: %v", err)
+			}
+			return ctx
+		}).Feature()
+
+	// test feature
+	testenv.Test(t, f)
+}
+...
+```
+
+The test steps are run in the order based on their level (Setup, Assess,
+Teardown).
+
+The example is available here: [e2e-framework POC]
+
+#### Blackbox testing
+
+The framework is using the controller-runtime client underneath, wrapped under
+its `klient`, which offers some abstractions to make testing easier. In order to
+deploy Crossplane resources, we would need to register Crossplane schemas within
+the client. Such a setup is more convenient for running integration tests. For
+end-to-end testing we should simulate users' interaction with the system and
+that happens very often (as advocated in Crossplane doc)typically using kubectl
+CLI.
```

Comment:
I would not over-index on needing to use `kubectl` to interact with the system as a black box.

I would argue that any client of the API server - even a programmatic client - is perfectly fine. You're still treating the system (the control plane) as a black box. I can see the value of testing that a really popular client (`kubectl`) works as we'd expect it. On the other hand we have plenty of control plane clients that are not `kubectl` (and are probably written with client-go) like ArgoCD, Helm, https://github.com/upbound/xgql, etc.

In my mind the fragility of shelling out to `kubectl`, `jq`, etc is not worth the benefit we get from testing that that client specifically.

Reply:
+1